Complete the exercises below For **Assignment #13**.

Load the `ISLR2` and the `tidymodels` packages.

In [1]:
library('ISLR2')
library('tidymodels')



── Attaching packages ────────────────────────────────────── tidymodels 1.2.0 ──

✔ broom        1.0.5      ✔ recipes      1.0.10
✔ dials        1.2.1      ✔ rsample      1.2.1 
✔ dplyr        1.1.4      ✔ tibble       3.2.1 
✔ ggplot2      3.5.1      ✔ tidyr        1.3.1 
✔ infer        1.0.7      ✔ tune         1.2.1 
✔ modeldata    1.3.0      ✔ workflows    1.1.4 
✔ parsnip      1.2.1      ✔ workflowsets 1.1.0 
✔ purrr        1.0.2      ✔ yardstick    1.3.1 

── Conflicts ───────────────────────────────────────── tidymodels_conflicts() ──
✖ purrr::discard() masks scales::discard()
✖ dplyr::filter()  masks stats::filter()
✖ dplyr::lag()     masks stats::lag()
✖ recipes::step()  masks stats::step()
• Learn how to get started at https://www.tidymodels.org/start/



In this assignment we will use the `Default` dataset which includes the default status for credit card customers (`default` variable) in addition to each customer's:

1. credit card balance (`balance` variable),
1. student status (`student` variable), and,
1. income (`income` variable).

In [2]:
Default |> head()

default student balance   income   
1 No      No       729.5265 44361.625
2 No      Yes      817.1804 12106.135
3 No      No      1073.5492 31767.139
4 No      No       529.2506 35704.494
5 No      No       785.6559 38463.496
6 No      Yes      919.5885  7491.559

We will be modeling `default` with the customer features.

Before we begin let's count how many customers fall into each `default` category.

In [3]:
Default |> count(default)

default n   
1 No      9667
2 Yes      333

The data is quite imbalanced. This will be important to keep in mind when we evaluate the performance of our model later. 

Run the code below to create and training data from `Default`. We will use the "test" dataset at the end to get a final evaluation of our best model's accuracy.

In [4]:
Default_split = initial_split(Default, prop = 0.90, strata = default)

Default_train = training(Default_split)
Default_test = testing(Default_split)

Create a logistic regression model called `mod`. Set the engine to `glm` and the mode to `classification`. 

In [28]:
mod <- 
    logistic_reg() |> 
    set_engine("glm") |> 
    set_mode("classification")


Our data is imbalanced. As such, a naive model that *always* predicts a customer to **not default** would be correct quite often. Let's start by calculating the "accuracy" of a naive model. This will be the baseline accuracy by which we evaluate other models.

In [6]:
# This code calculates the accuracy of a model that always predicts default to be "No"

Default_train |>
    mutate(.pred_naive = factor('No', levels = c('No', 'Yes'))) |>
    accuracy(truth = default, .pred_naive)

.metric  .estimator .estimate
1 accuracy binary     0.967

Let's use k-fold cross validation to evaluate the performance of a model where the outcome is `default` and the predictors are `income` and `balance`.

To start, use `vfold_cv` to generate 10 validation folds (i.e. set the `v` variable to 10). Set the `strata` argument to `default` so we preserve the distribution of `default` values in each fold.

Creat your folds below and use `glimpse` to look at the output table. Call your output folds tables "folds".

In [8]:
folds <-  vfold_cv(data = Default, v = 10, strata = default)

folds |> glimpse()

Rows: 10
Columns: 2
$ splits <list> [<vfold_split[9000 x 1000 x 10000 x 4]>], [<vfold_split[9000 x…
$ id     <chr> "Fold01", "Fold02", "Fold03", "Fold04", "Fold05", "Fold06", "Fo…


The code below fits a model to each of your 10 folds. `collect_metrics` finds the average of evaluation metrics for each of your ten models. 

In [9]:
mod |> 
    fit_resamples(default ~ income + balance, folds) |>
    collect_metrics()

.metric     .estimator mean       n  std_err      .config             
1 accuracy    binary     0.97370000 10 0.0011160446 Preprocessor1_Model1
2 brier_class binary     0.02147883 10 0.0007922781 Preprocessor1_Model1
3 roc_auc     binary     0.94970283 10 0.0050654134 Preprocessor1_Model1

❓How does the model accuracy compare to the naive model from above?

The accuracy of the naive model above was 0.967.  
The accuracy of the model averaged over the 10 folds was 0.97370000.
This is a difference of 0.00670  
If we make a 95% confidence interval for the accuracy of our model.
.9737 ± 1.96 * 0.0011160446
(0.9715, 0.9759)  
we can see that our model appears better than the naive model.

Complete the cell below to evaluate a model also includes the `student` variable as as predictor.
1. use `default ~ income + balance + student` as the formula,
2. encode your `student` variable with `step_dummy`, and,
3. don't forget to `prep` your recipe!

In [23]:
rec <- recipe(default ~ income + balance + student, data = Default) |> 
    step_dummy(student)

mod |>
    fit_resamples(rec, folds) |>
    collect_metrics()

.metric     .estimator mean      n  std_err      .config             
1 accuracy    binary     0.9733000 10 0.0010959521 Preprocessor1_Model1
2 brier_class binary     0.0214026 10 0.0007464561 Preprocessor1_Model1
3 roc_auc     binary     0.9500049 10 0.0051208588 Preprocessor1_Model1

❓Does it appear that the model that includes `student` improves upon the first model with only `income` and `balance` as predictors?

No. The accuracy is slightly lower with a very similar std_err.



Finally, estimate the accuracy of an `default ~ income + balance` model on the test data, `Default_test`. 

❓Does our model outperform a naive model?

In [32]:
mod_fit <- mod |> 
    fit(default ~ income + balance, data = Default_train)

predictions <- predict(mod_fit, new_data = Default_test, type = "class") |>
    bind_cols(Default_test)

In [33]:
predictions |> head()


.pred_class default student balance   income  
1 No          No      No      1161.0579 37468.53
2 No          No      No       237.0451 28251.70
3 No          No      No         0.0000 50265.31
4 No          No      No       485.9369 61566.11
5 No          No      No       228.9525 50500.18
6 No          No      No       773.2117 34353.31

In [37]:
accuracy <- mean(predictions$.pred_class == Default_test$default)
accuracy

[1] 0.966

The mdoel does not outperform a naive model.